// simply-mcp/dist/src/api/functional/builders.js
/**
 * Functional API Builders
 *
 * Builder functions and classes for creating MCP server configurations.
 *
 * Inspired by FastMCP - allows defining MCP servers using a declarative configuration
 * with tools, prompts, and resources.
 *
 * @example
 * ```typescript
 * import { defineMCP } from 'simply-mcp';
 * import { z } from 'zod';
 *
 * export default defineMCP({
 *   name: 'my-server',
 *   version: '1.0.0',
 *   tools: [
 *     {
 *       name: 'greet',
 *       description: 'Greet a user',
 *       parameters: z.object({ name: z.string() }),
 *       execute: async (args) => `Hello, ${args.name}!`
 *     }
 *   ]
 * });
 * ```
 */
import { Schema } from '../../schema-builder.js';
// Re-export Schema builder for convenience
export { Schema };
/**
 * Helper function to define an MCP configuration with type safety
 * This function doesn't do anything at runtime - it just provides type checking
 */
export function defineMCP(config) {
    return config;
}
/**
 * Helper function to define a tool with type safety
 */
export function defineTool(tool) {
    return tool;
}
/**
 * Helper function to define a prompt with type safety
 */
export function definePrompt(prompt) {
    return prompt;
}
/**
 * Helper function to define a resource with type safety
 */
export function defineResource(resource) {
    return resource;
}
/**
 * Builder pattern for creating MCP configs (alternative to defineMCP)
 * Inspired by FastMCP's decorator pattern
 */
export class MCPBuilder {
    config;
    constructor(options) {
        this.config = {
            name: options.name,
            version: options.version,
            port: options.port,
            basePath: options.basePath,
            defaultTimeout: options.defaultTimeout,
            tools: [],
            prompts: [],
            resources: [],
        };
    }
    /**
     * Add a tool to the server
     */
    tool(tool) {
        this.config.tools = this.config.tools || [];
        this.config.tools.push(tool);
        return this;
    }
    /**
     * Add a prompt to the server
     */
    prompt(prompt) {
        this.config.prompts = this.config.prompts || [];
        this.config.prompts.push(prompt);
        return this;
    }
    /**
     * Add a resource to the server
     */
    resource(resource) {
        this.config.resources = this.config.resources || [];
        this.config.resources.push(resource);
        return this;
    }
    /**
     * Build and return the config
     */
    build() {
        return this.config;
    }
}
/**
 * Create a new MCP builder
 */
export function createMCP(options) {
    return new MCPBuilder(options);
}
