// simply-mcp/dist/src/index.js
/**
 * SimplyMCP - Model Context Protocol Server Framework
 *
 * A comprehensive MCP server framework with support for multiple API styles:
 * - Decorator-based class API
 * - Functional single-file API
 * - Programmatic API
 * - Interface-driven API (v2.5.0+)
 * - MCP Builder API (v2.5.0+) - Build MCP servers using MCP itself
 *
 * @example Decorator API
 * ```typescript
 * import { MCPServer } from 'simply-mcp';
 *
 * @MCPServer({ name: 'my-server', version: '1.0.0' })
 * class MyServer {
 *   greet(name: string): string {
 *     return `Hello, ${name}!`;
 *   }
 * }
 * ```
 *
 * @example Functional API
 * ```typescript
 * import { defineMCP } from 'simply-mcp';
 * import { z } from 'zod';
 *
 * export default defineMCP({
 *   name: 'my-server',
 *   version: '1.0.0',
 *   tools: [{
 *     name: 'greet',
 *     description: 'Greet a user',
 *     parameters: z.object({ name: z.string() }),
 *     execute: async (args) => `Hello, ${args.name}!`
 *   }]
 * });
 * ```
 *
 * @example Programmatic API
 * ```typescript
 * import { BuildMCPServer } from 'simply-mcp';
 * import { z } from 'zod';
 *
 * const server = new BuildMCPServer({ name: 'my-server', version: '1.0.0' });
 * server.addTool({
 *   name: 'greet',
 *   description: 'Greet a user',
 *   parameters: z.object({ name: z.string() }),
 *   execute: async (args) => `Hello, ${args.name}!`
 * });
 * await server.start();
 * ```
 *
 * @example Interface-Driven API (v3.0.0 - In Development)
 * ```typescript
 * import type { ITool, IServer } from 'simply-mcp';
 *
 * interface GreetTool extends ITool {
 *   name: 'greet';
 *   description: 'Greet a user';
 *   params: { name: string };
 *   result: string;
 * }
 *
 * interface MyServerInterface extends IServer {
 *   name: 'my-server';
 *   version: '1.0.0';
 * }
 *
 * export default class MyServer implements MyServerInterface {
 *   greet: GreetTool = async (params) => `Hello, ${params.name}!`;
 * }
 * ```
 *
 * @example MCP Builder API (v2.5.0+)
 * ```typescript
 * import { defineMCPBuilder, ValidationToolsPreset, WorkflowPromptsPreset } from 'simply-mcp';
 *
 * export default defineMCPBuilder({
 *   name: 'mcp-dev',
 *   version: '1.0.0',
 *   toolPresets: [ValidationToolsPreset],
 *   promptPresets: [WorkflowPromptsPreset]
 * });
 * ```
 */
// ============================================================================
// Main SimplyMCP Class (Programmatic API - Legacy)
// ============================================================================
/**
 * @deprecated Use `BuildMCPServer` instead (renamed in v2.5.0+)
 * This export will be removed in v3.0.0.
 *
 * @example
 * ```typescript
 * // NEW: Use BuildMCPServer
 * import { BuildMCPServer } from 'simply-mcp';
 * const server = new BuildMCPServer({ name: 'my-server', version: '1.0.0' });
 *
 * // OLD: SimplyMCP (deprecated)
 * import { SimplyMCP } from 'simply-mcp';
 * const server = new SimplyMCP({ name: 'my-server', version: '1.0.0' });
 * ```
 */
export { SimplyMCP } from './SimplyMCP.js';
// ============================================================================
// BuildMCPServer Class (Programmatic API - New)
// ============================================================================
// New programmatic API with improved naming
export { BuildMCPServer } from './api/programmatic/BuildMCPServer.js';
// ============================================================================
// Decorator API
// ============================================================================
export { MCPServer, tool, prompt, resource, } from './decorators.js';
// ============================================================================
// Single-File Functional API
// ============================================================================
export { defineMCP, defineTool, definePrompt, defineResource, createMCP, MCPBuilder, Schema, } from './single-file-types.js';
// ============================================================================
// Schema Builder
// ============================================================================
export { schemaToZod, } from './schema-builder.js';
export { parseInterfaceFile, snakeToCamel, loadInterfaceServer, isInterfaceFile, } from './api/interface/index.js';
// ============================================================================
// MCP Builder API (v2.5.0+)
// ============================================================================
/**
 * MCP Builder API - Build MCP servers using MCP itself
 *
 * A specialized API for creating MCP servers that help build other MCP servers.
 * Uses sampling-based validation to leverage your LLM for intelligent feedback.
 *
 * Layer 1: Foundation - Basic design tools
 * Layer 2: Feature - Sampling-based validation, workflow guidance
 *
 * @example Basic usage
 * ```typescript
 * import { defineMCPBuilder, DesignToolsPreset } from 'simply-mcp';
 *
 * export default defineMCPBuilder({
 *   name: 'mcp-dev',
 *   version: '1.0.0',
 *   toolPresets: [DesignToolsPreset]
 * });
 * ```
 *
 * @example Advanced usage with sampling
 * ```typescript
 * import {
 *   createMCPBuilder,
 *   ValidationToolsPreset,
 *   WorkflowPromptsPreset
 * } from 'simply-mcp';
 *
 * export default createMCPBuilder({
 *   name: 'mcp-dev-complete',
 *   version: '1.0.0'
 * })
 *   .useToolPreset(ValidationToolsPreset)
 *   .usePromptPreset(WorkflowPromptsPreset)
 *   .build();
 * ```
 */
// Builder functions
export { defineMCPBuilder, createMCPBuilder, MCPBuilderBuilder, } from './api/mcp/builders.js';
// Adapter functions
export { createServerFromMCPBuilder, loadMCPBuilderServer, isMCPBuilderFile, } from './api/mcp/adapter.js';
// Presets
export { DesignToolsPreset, // Layer 1: Basic design tools
ValidationToolsPreset, // Layer 2: Sampling-based validation
InteractiveValidationToolsPreset, // Layer 2: Interactive validation (no sampling - works with Claude Code CLI!)
CodeGenerationToolsPreset, // Layer 2: Code generation (complete the workflow!)
WorkflowPromptsPreset, // Layer 2: Workflow guidance
 } from './api/mcp/presets/index.js';
// ============================================================================
// Error Classes
// ============================================================================
export { HandlerExecutionError, HandlerLoadError, HandlerSyntaxError, HandlerTimeoutError, HandlerNetworkError, HandlerNotFoundError, HandlerConfigError, HandlerPermissionError, } from './core/errors.js';
export { defineConfig } from './config.js';
