// simply-mcp/dist/src/handlers/RegistryHandlerResolver.js
/**
 * Registry Handler Resolver - Looks up pre-registered handler functions
 */
import { HandlerNotFoundError, HandlerConfigError } from '../core/errors.js';
/**
 * Resolver for registry-based handlers
 * Maintains an in-memory registry of pre-registered handler functions
 */
export class RegistryHandlerResolver {
    registry = new Map();
    /**
     * Check if this resolver can handle the given configuration
     */
    canResolve(config) {
        return config.type === 'registry';
    }
    /**
     * Resolve a registry handler configuration to an executable function
     */
    async resolve(config) {
        if (!this.canResolve(config)) {
            throw new HandlerConfigError(`RegistryHandlerResolver cannot resolve handler of type: ${config.type}`);
        }
        const registryConfig = config;
        const handler = this.registry.get(registryConfig.name);
        if (!handler) {
            const availableHandlers = Array.from(this.registry.keys()).join(', ');
            throw new HandlerNotFoundError(`Handler '${registryConfig.name}' not found in registry. Available handlers: ${availableHandlers || 'none'}`, { name: registryConfig.name, availableHandlers });
        }
        return handler;
    }
    /**
     * Register a handler function in the registry
     * @param name Name to register the handler under
     * @param handler Handler function to register
     */
    register(name, handler) {
        if (this.registry.has(name)) {
            throw new HandlerConfigError(`Handler '${name}' is already registered. Use unregister() first if you want to replace it.`, { name });
        }
        this.registry.set(name, handler);
    }
    /**
     * Unregister a handler from the registry
     * @param name Name of the handler to unregister
     * @returns True if handler was found and removed, false otherwise
     */
    unregister(name) {
        return this.registry.delete(name);
    }
    /**
     * Check if a handler is registered
     * @param name Name of the handler to check
     * @returns True if handler is registered
     */
    has(name) {
        return this.registry.has(name);
    }
    /**
     * Get all registered handler names
     * @returns Array of handler names
     */
    list() {
        return Array.from(this.registry.keys());
    }
    /**
     * Clear all registered handlers
     */
    clear() {
        this.registry.clear();
    }
    /**
     * Get the number of registered handlers
     * @returns Number of handlers in registry
     */
    size() {
        return this.registry.size;
    }
}
