// simply-mcp/src/core/index.ts
/**
 * Core exports for the Handler Execution Framework
 */

export * from './types.js';
export * from './errors.js';
export * from './logger.js';
export * from './HandlerManager.js';

// Export content helpers runtime functions
export {
  createImageContent,
  createAudioContent,
  createBlobContent,
} from './content-helpers.js';

// Export content helper types separately (type-only)
export type {
  ImageInput,
  AudioInput,
  BinaryInput,
} from './content-helpers.js';

// Inline dependency management (Phase 2, Feature 2)
export * from '../features/dependencies/dependency-types';
export * from '../features/dependencies/dependency-parser';
export * from '../features/dependencies/dependency-validator';

// Export dependency-utils, avoiding conflicts with dependency-resolver
export {
  mergeDependencies,
  filterDependencies,
} from '../features/dependencies/dependency-utils';

// Auto-installation (Phase 2, Feature 3)
export * from '../features/dependencies/installation-types';
export * from '../features/dependencies/dependency-installer';
export * from '../features/dependencies/dependency-checker';
export * from '../features/dependencies/package-manager-detector';

// Bundling (Phase 2, Feature 4)
export * from '../features/dependencies/bundle-types';
export * from './bundler.js';
export * from './entry-detector.js';

// Export dependency-resolver exports except the conflicting ones
export {
  resolveDependencies,
} from '../features/dependencies/dependency-resolver';

export * from './output-formatter.js';
