// simply-mcp/dist/src/core/logger.js
/**
 * Logger implementation for handler execution
 */
/**
 * Console-based logger implementation with optional MCP notification support
 */
export class ConsoleLogger {
    prefix;
    notificationCallback;
    silent;
    constructor(prefix = '[Handler]', notificationCallback, silent = false) {
        this.prefix = prefix;
        this.notificationCallback = notificationCallback;
        this.silent = silent;
    }
    debug(message, ...args) {
        if (!this.silent) {
            console.error(`${this.prefix} [DEBUG]`, message, ...args);
        }
        this.sendNotification('debug', message, args);
    }
    info(message, ...args) {
        if (!this.silent) {
            console.error(`${this.prefix} [INFO]`, message, ...args);
        }
        this.sendNotification('info', message, args);
    }
    warn(message, ...args) {
        if (!this.silent) {
            console.warn(`${this.prefix} [WARN]`, message, ...args);
        }
        this.sendNotification('warning', message, args);
    }
    error(message, ...args) {
        if (!this.silent) {
            console.error(`${this.prefix} [ERROR]`, message, ...args);
        }
        this.sendNotification('error', message, args);
    }
    notice(message, ...args) {
        if (!this.silent) {
            console.error(`${this.prefix} [NOTICE]`, message, ...args);
        }
        this.sendNotification('notice', message, args);
    }
    critical(message, ...args) {
        if (!this.silent) {
            console.error(`${this.prefix} [CRITICAL]`, message, ...args);
        }
        this.sendNotification('critical', message, args);
    }
    alert(message, ...args) {
        if (!this.silent) {
            console.error(`${this.prefix} [ALERT]`, message, ...args);
        }
        this.sendNotification('alert', message, args);
    }
    emergency(message, ...args) {
        if (!this.silent) {
            console.error(`${this.prefix} [EMERGENCY]`, message, ...args);
        }
        this.sendNotification('emergency', message, args);
    }
    sendNotification(level, message, args) {
        if (this.notificationCallback) {
            const data = args && args.length > 0 ? args : undefined;
            this.notificationCallback(level, message, data);
        }
    }
}
/**
 * Create a default logger instance
 */
export function createDefaultLogger(prefix, notificationCallback, silent) {
    return new ConsoleLogger(prefix, notificationCallback, silent);
}
