// simply-mcp/dist/src/api/interface/InterfaceServer.js
/**
 * InterfaceServer - Wrapper for BuildMCPServer with MCP Protocol Methods
 *
 * Bridges interface-based server definitions to the MCP protocol.
 * Provides public methods for MCP operations: listTools, executeTool, listPrompts, etc.
 */
/**
 * InterfaceServer - MCP Protocol Wrapper
 *
 * Wraps BuildMCPServer to provide public MCP protocol methods.
 * This allows interface-driven servers to expose the same API as class-based servers.
 *
 * @example
 * ```typescript
 * import { loadInterfaceServer } from 'simply-mcp';
 *
 * const server = await loadInterfaceServer({ filePath: './server.ts' });
 *
 * // List all tools
 * const tools = server.listTools();
 *
 * // Execute a tool
 * const result = await server.executeTool('greet', { name: 'Alice' });
 *
 * // Read a resource
 * const data = await server.readResource('config://server');
 *
 * // Start the server
 * await server.start();
 * ```
 */
export class InterfaceServer {
    buildServer;
    /**
     * Create a new InterfaceServer wrapper
     * @param buildServer The underlying BuildMCPServer instance
     */
    constructor(buildServer) {
        this.buildServer = buildServer;
    }
    // ===== Metadata Getters =====
    /**
     * Get the server name
     */
    get name() {
        return this.buildServer.name;
    }
    /**
     * Get the server version
     */
    get version() {
        return this.buildServer.version;
    }
    /**
     * Get the server description
     */
    get description() {
        return this.buildServer.description;
    }
    // ===== MCP Protocol Methods - Tools =====
    /**
     * List all registered tools
     * @returns Array of tool definitions with name, description, and input schema
     */
    listTools() {
        const tools = this.buildServer.getTools();
        return Array.from(tools.values()).map((tool) => ({
            name: tool.definition.name,
            description: tool.definition.description,
            inputSchema: tool.jsonSchema,
        }));
    }
    /**
     * Execute a tool by name
     * @param toolName Name of the tool to execute
     * @param args Arguments for the tool
     * @returns Tool execution result
     */
    async executeTool(toolName, args) {
        return await this.buildServer.executeToolDirect(toolName, args);
    }
    // ===== MCP Protocol Methods - Prompts =====
    /**
     * List all registered prompts
     * @returns Array of prompt definitions with name, description, and arguments
     */
    listPrompts() {
        const prompts = this.buildServer.getPrompts();
        return Array.from(prompts.values()).map((prompt) => ({
            name: prompt.name,
            description: prompt.description,
            arguments: prompt.arguments,
        }));
    }
    /**
     * Get a prompt by name with rendered template
     * @param promptName Name of the prompt
     * @param args Arguments for template interpolation
     * @returns Prompt result with rendered messages
     */
    async getPrompt(promptName, args = {}) {
        return await this.buildServer.getPromptDirect(promptName, args);
    }
    // ===== MCP Protocol Methods - Resources =====
    /**
     * List all registered resources
     * @returns Array of resource definitions with uri, name, description, and mimeType
     */
    listResources() {
        const resources = this.buildServer.getResources();
        return Array.from(resources.values()).map((resource) => ({
            uri: resource.uri,
            name: resource.name,
            description: resource.description,
            mimeType: resource.mimeType,
        }));
    }
    /**
     * Read a resource by URI
     * @param uri URI of the resource to read
     * @returns Resource contents
     */
    async readResource(uri) {
        return await this.buildServer.readResourceDirect(uri);
    }
    // ===== Lifecycle Methods =====
    /**
     * Start the MCP server
     * @param options Start options (transport, port, stateful)
     */
    async start(options) {
        await this.buildServer.start(options);
    }
    /**
     * Stop the MCP server gracefully
     */
    async stop() {
        await this.buildServer.stop();
    }
    /**
     * Get server information
     * @returns Server info with name, version, and running status
     */
    getInfo() {
        return this.buildServer.getInfo();
    }
    /**
     * Get statistics about registered items
     * @returns Stats with counts of tools, prompts, and resources
     */
    getStats() {
        return this.buildServer.getStats();
    }
    // ===== Direct Access to BuildMCPServer =====
    /**
     * Get the underlying BuildMCPServer instance
     * For advanced use cases that need direct access
     * @returns The wrapped BuildMCPServer instance
     */
    getBuildServer() {
        return this.buildServer;
    }
}
